### PR TITLE
Add field-level warning plumbing and add pod spec warnings

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

* Sets up plumbing for REST handlers to add object-specific warnings on create or update
* Adds warnings for known-bad values in pod-specs

#### Which issue(s) this PR fixes:
xref https://github.com/kubernetes/kubernetes/issues/94626

xref kubernetes/enhancements#1693

#### Does this PR introduce a user-facing change?
```release-note
Warnings for use of deprecated and known-bad values in pod specs are now sent
```

### CODE REVIEW

File: pkg/registry/core/pod/strategy.go [hunk 1]
```diff
@@ -130,6 +132,7 @@ func (podStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object)
 
 // WarningsOnUpdate returns warnings for the given update.
 func (podStrategy) WarningsOnUpdate(ctx context.Context, obj, old runtime.Object) []string {
+	// skip warnings on pod update, since the fields we warn about are generally immutable on pods
```
Comment (smarterclayton) [inline pkg/registry/core/pod/strategy.go:135]: i predict this comment will not age well

Comment (liggitt) [reply pkg/registry/core/pod/strategy.go:135]: and since humans don't generally interact directly with pods

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 2]
```diff
@@ -51,6 +52,19 @@ type RESTUpdateStrategy interface {
 	// filled in before the object is persisted.  This method should not mutate
 	// the object.
 	ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList
+	// WarningsOnUpdate returns any warnings that should be returned to the updater.
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:55]: Reads a bit wierdly.  `WarningsOnUpdate returns warnings to the client performing an update.`

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 3]
```diff
@@ -51,6 +52,19 @@ type RESTUpdateStrategy interface {
...
+	// WarningsOnUpdate is invoked after default fields in the object have been filled in
+	// and after ValidateUpdate has passed, before Canonicalize is called, and before the object is persisted.
+	// This method must not mutate either object.
+	//
+	// Be brief; limit warnings to 120 characters if possible.
+	// Don't include a "Warning:" prefix in the message (that is added by clients on output).
+	// Warnings returned about a specific field should be formatted as "path.to.field: message".
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:62]: Maybe given an example so people don't do `pod.spec.foo` like they did in validation?

File: staging/src/k8s.io/apiserver/pkg/registry/rest/update.go [hunk 4]
```diff
@@ -51,6 +52,19 @@ type RESTUpdateStrategy interface {
...
+	//
+	// Use warning messages to describe problems the client making the API request should correct or be aware of.
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/update.go:64]: What other guidance can we give?  Maybe examples of how you plan to use it (2-3)?

File: pkg/registry/batch/cronjob/strategy.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ import (
 	"context"
```
Comment (smarterclayton) [inline pkg/registry/batch/cronjob/strategy.go:20]: Not sure if you want to pull this out into another PR or not, but would merge faster.

File: pkg/registry/batch/cronjob/strategy.go [hunk 6]
```diff
@@ -20,6 +20,7 @@ import (
...
 
 	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
```
Comment (smarterclayton) [inline pkg/registry/batch/cronjob/strategy.go:23]: Should there be an integration or REST test that enforces this?

File: pkg/api/pod/warnings.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+	},
+}
+
+func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta *metav1.ObjectMeta, oldPodSpec *api.PodSpec, oldMeta *metav1.ObjectMeta) []string {
+	var warnings []string
+
+	// use of deprecated node labels in selectors/affinity/topology
+	for k := range podSpec.NodeSelector {
+		if msg, deprecated := deprecatedNodeLabels[k]; deprecated {
+			warnings = append(warnings, fieldPath.Child("spec", "nodeSelector").Key(k).String()+": "+msg)
```
Comment (smarterclayton) [inline pkg/api/pod/warnings.go:101]: use a fmt.Sprintf istead of +

File: pkg/api/pod/warnings.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+		for i, item := range podSpec.HostAliases {
+			if items.Has(item.IP) {
+				warnings = append(warnings, fmt.Sprintf("%s: duplicate ip %q", fieldPath.Child("spec", "hostAliases").Index(i).Child("ip"), item.IP))
+			} else {
+				items.Insert(item.IP)
+			}
+		}
+	}
+
+	// duplicate imagePullSecrets (#91629, #58477)
```
Comment (smarterclayton) [inline pkg/api/pod/warnings.go:172]: I admit i teared up a bit when i saw you fixing bugs with warnings.  Beautiful.

File: pkg/api/pod/warnings.go (hunk 7 above)
Comment (liggitt) [reply pkg/api/pod/warnings.go:101]: really?

File: pkg/api/pod/warnings.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	// imagePullSecrets with empty name (#99454#issuecomment-787838112)
+	for i, item := range podSpec.ImagePullSecrets {
+		if len(item.Name) == 0 {
+			warnings = append(warnings, fmt.Sprintf("%s: invalid empty name %q", fieldPath.Child("spec", "imagePullSecrets").Index(i).Child("name"), item.Name))
+		}
+	}
+
+	// duplicate volume names (#78266, #58477)
+	if len(podSpec.Volumes) > 1 {
+		items := sets.NewString()
```
Comment (smarterclayton) [inline pkg/api/pod/warnings.go:192]: This will allocate a bunch.  Not the end of the world, but is someting to keep in mind

File: pkg/api/pod/warnings.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+		if value, ok := c.Resources.Limits[api.ResourceEphemeralStorage]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "limits").Key(string(api.ResourceEphemeralStorage)), value.String()))
+		}
+		if value, ok := c.Resources.Requests[api.ResourceEphemeralStorage]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "requests").Key(string(api.ResourceEphemeralStorage)), value.String()))
+		}
+
+		// duplicate containers[*].env (#86163, #93266, #58477)
+		if len(c.Env) > 1 {
+			items := sets.NewString()
```
Comment (smarterclayton) [inline pkg/api/pod/warnings.go:227]: While it would be uglier, it would be better to reuse one string set on this entire function and reset before use.

Comment (liggitt) [reply pkg/api/pod/warnings.go:227]: how much better? benchmarks didn't look that concerning to me

```
BenchmarkNoWarnings-12    	  491527	      2534 ns/op	     336 B/op	       7 allocs/op
```

clearing and reusing that set between blocks actually made the benchmark worse:

```
BenchmarkNoWarnings-12    	  521125	      2552 ns/op	     528 B/op	       9 allocs/op
```

Comment (smarterclayton) [reply pkg/api/pod/warnings.go:227]: I don't know how clearing and reusing could make it worse - how did you clear?

File: pkg/api/pod/warnings.go (hunk 7 above)
Comment (smarterclayton) [reply pkg/api/pod/warnings.go:101]: yes!

File: pkg/api/pod/warnings.go (hunk 10 above)
Comment (smarterclayton) [reply pkg/api/pod/warnings.go:227]: This is a bug in the set - it should have a Clear() or Reset() method that does `for k := range m { delete(m, k) }` which the go compiler optimizes to "clear but reuse buckets", which is roughly free

Comment (liggitt) [reply pkg/api/pod/warnings.go:227]: hmm... reworked my clearing, dropped it back down to the original, but no cheaper than a new set per block:

```diff
+func clear(items sets.String) {
+       for k := range items {
+               delete(items, k)
+       }
+}
+
 func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta *metav1.ObjectMeta, oldPodSpec *api.PodSpec, oldMeta
        var warnings []string
+       var items sets.String

        // use of deprecated node labels in selectors/affinity/topology
        for k := range podSpec.NodeSelector {
@@ -159,7 +166,11 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta

        // duplicate hostAliases (#91670, #58477)
        if len(podSpec.HostAliases) > 1 {
-               items := sets.NewString()
+               if items == nil {
+                       items = sets.NewString()
+               } else {
+                       clear(items)
+               }
```

```
BenchmarkNoWarnings-12    	  702271	      1959 ns/op	     336 B/op	       7 allocs/op
```

File: pkg/api/pod/warnings.go (hunk 7 above)
Comment (liggitt) [reply pkg/api/pod/warnings.go:101]: but... https://gist.github.com/dtjm/c6ebc86abe7515c988ec

File: pkg/api/pod/warnings.go (hunk 9 above)
Comment (liggitt) [reply pkg/api/pod/warnings.go:192]: benchmarks don't show it allocating more than creating a single string set and clearing/reusing it

File: pkg/api/pod/warnings.go (hunk 10 above)
Comment (liggitt) [reply pkg/api/pod/warnings.go:227]: is there another way I should be benchmarking to demonstrate the value of the increased ugliness?

File: pkg/registry/batch/cronjob/strategy.go (hunk 5 above)
Comment (liggitt) [reply pkg/registry/batch/cronjob/strategy.go:20]: eh... if this drags on, I might, but hopefully it won't

File: pkg/registry/batch/cronjob/strategy.go (hunk 6 above)
Comment (liggitt) [reply pkg/registry/batch/cronjob/strategy.go:23]: added to unit tests of strategy

**Comment by liggitt** [comment]
all comments except https://github.com/kubernetes/kubernetes/pull/101688#discussion_r631374505 addressed (since I can't see an improvement in the benchmark), diff in https://github.com/kubernetes/kubernetes/compare/0b6e593a578f77715d28283f0c95257905624d40..f669796dfdc001c4426751a419c0fc29b5b66d60

File: pkg/api/pod/warnings.go (hunk 10 above)
Comment (smarterclayton) [reply pkg/api/pod/warnings.go:244]: Are you comfortable with me getting a rain check to come back later and frown at the impact it has and volunteering to review any slight ugliness I add later?  I'm comfortable deferring since micro didn't improve.
